src/Dhall.hs

```diff
@@ -529,104 +544,88 @@ instance (Constructor c1, Constructor c2, GenericInterpret f1, GenericInterpret
 
-        extract u@(UnionLit name' e _)
-            | name == name' = fmap (L1 . M1) (extractL e)
-            | otherwise     = fmap  R1       (extractR u)
-        extract _ = Nothing
-
-        expected = Union (Data.Map.insert name expectedL expectedR)
-
-        Type extractL        expectedL  = genericAutoWith options
-        Type extractR (Union expectedR) = genericAutoWith options
-
 instance (GenericInterpret (f :+: g), GenericInterpret (h :+: i)) => GenericInterpret ((f :+: g) :+: (h :+: i)) where
-    genericAutoWith options = Type {..}
-      where
-        extract e = fmap L1 (extractL e) <|> fmap R1 (extractR e)
-
-        expected = Union (Data.Map.union expectedL expectedR)
-
-        Type extractL (Union expectedL) = genericAutoWith options
-        Type extractR (Union expectedR) = genericAutoWith options
+    genericAutoWith options = do
+        Type extractL (Union expectedL) <- genericAutoWith options
+        Type extractR (Union expectedR) <- genericAutoWith options
+        let extract e = fmap L1 (extractL e) <|> fmap R1 (extractR e)
+        let expected = Union (Data.Map.union expectedL expectedR)
+        pure (Type {..})
 
 instance GenericInterpret f => GenericInterpret (M1 C c f) where
-    genericAutoWith = fmap (fmap M1) genericAutoWith
+    genericAutoWith options = do
+        res <- genericAutoWith options
+        pure (fmap M1 res)
 
 instance GenericInterpret U1 where
-    genericAutoWith _ = Type {..}
+    genericAutoWith _ = pure (Type {..})
       where
         extract _ = Just U1
 
         expected = Record (Data.Map.fromList [])
 
 instance (GenericInterpret f, GenericInterpret g) => GenericInterpret (f :*: g) where
-    genericAutoWith options = Type {..}
-      where
-        extract = liftA2 (liftA2 (:*:)) extractL extractR
-
-        expected = Record (Data.Map.union ktsL ktsR)
-          where
-            Record ktsL = expectedL
-            Record ktsR = expectedR
-        Type extractL expectedL = genericAutoWith options
-        Type extractR expectedR = genericAutoWith options
+    genericAutoWith options = do
+        Type extractL expectedL <- genericAutoWith options
+        Type extractR expectedR <- genericAutoWith options
+        let Record ktsL = expectedL
+        let Record ktsR = expectedR
+        pure (Type { extract = liftA2 (liftA2 (:*:)) extractL extractR
+                        , expected = Record (Data.Map.union ktsL ktsR) })
 
 instance (Selector s, Interpret a) => GenericInterpret (M1 S s (K1 i a)) where
-    genericAutoWith opts@(InterpretOptions {..}) = Type {..}
+    genericAutoWith opts@(InterpretOptions {..}) = do
+        name <- case selName n of
+                    "" -> do i <- get
+                             put (i + 1)
+                             pure ("_" ++ show i)
+                    nn -> pure nn
+        pure (Type { expected = expected_name name, extract = extract_name name })
```

Comment:
Minor suggestion: you can simplify this a little bit by relocating `expected` and `extract` inside the `do` block, like this:

```haskell
name <- case ...
let extract (RecordLit m) = ...
let expected = ...
pure (Type {..})
```